src/ConnectionPool.jl

```diff
@@ -358,115 +383,45 @@ function purge(c::Connection)
     @ensure bytesavailable(c) == 0
 end
 
-"""
-The `pool` is a collection of open `Connection`s.  The `request`
-function calls `getconnection` to retrieve a connection from the
-`pool`.  When the `request` function has written a Request Message
-it calls `closewrite` to signal that the `Connection` can be reused
-for writing (to send the next Request). When the `request` function
-has read the Response Message it calls `closeread` to signal that
-the `Connection` can be reused for reading.
-"""
-const pool = [Vector{Connection}()]
-const poollock = [ReentrantLock()]
-const poolcondition = [Condition()]
-
 """
     closeall()
 
 Close all connections in `pool`.
 """
 function closeall()
-    lock(poollock[Threads.threadid()]) do
-        for c in pool[Threads.threadid()]
-            close(c)
+    for pod in values(POOL.conns)
+        while isready(pod.conns)
```

Comment:
this is race-y (you should be holding the `pod.conns` lock)

Reply:
is it common to lock a channel when calling `isready` and then `take!`? it seems like having a `maybetake!` might be convenient for the use-case of wanting an unblocked `take!` that returns `nothing` if there isn't items available.

Reply:
Updated